Library/Homebrew/rubocops/components_order.rb

```diff
@@ -24,10 +24,10 @@ class ComponentsOrder < FormulaCop
           [{ name: :version_scheme, type: :method_call }],
           [{ name: :head,      type: :method_call }],
           [{ name: :stable,    type: :block_call }],
+          [{ name: :head,      type: :block_call }],
           [{ name: :livecheck, type: :block_call }],
```

Comment:
Could `livecheck` move below `bottle` instead?

Reply:
The only technical requirement to make this work is that the `livecheck` block needs to come after `head`. Currently it comes after the method call version of `head` but not the block format.

I would prefer to relocate the `head` block because:

* It moves the `head` block near to the location of its method call, which is consistent with the `stable` block.
* It keeps the `homepage`, `stable`, and `head` URLs near to each other. Currently the `head` method call is in the first group (with `homepage`, `stable`, etc.) but the `head` block is separated by coming after `bottle` (instead of before `bottle`, like the `stable` block).
* It allows us to keep the `livecheck` block near the `homepage`, `stable`, and `head` URLs (which are used in checks). This increases the chances that a maintainer/contributor will modify a `livecheck` block to keep it up to date when they modify one of these URLs. If we move the `livecheck` block below the current `head` block location, then it's easier to overlook.

---

I think there are clear benefits to relocating the `head` block but I don't see the benefits of maintaining the current location (after `bottle`) and moving the `livecheck` block below it. If the goal of the current `head` block location is to keep the `depends_on`/`patch`/`resource` statements closer together, then moving the `livecheck` block after the `head` block's current location disrupts this anyway.

Additionally, `stable` blocks can contain `patch`/`resource` blocks (e.g., [`agda`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/agda.rb)), so this already does what we may be trying to avoid with the current `head` block location. Having `stable` and `head` straddle the `bottle` block feels unusual to me and it seems more logical to keep them together and near to their original method calls (before the bottle block).

Is this reasoning compelling or are we 100% set on not moving the `head` block for any reason?

Reply:
> Is this reasoning compelling or are we 100% set on not moving the `head` block for any reason?

I'd rather avoid moving it for the sake of the `livecheck` block. I think the `bottle` block is more "important"/commonly read than `head` (or even `livecheck`). 

If the issue is "the livecheck block relies on other DSLs that need to be declared before it" I think it should be moved down rather than others moved up.

Reply:
One benefit to moving the bottle block downward is that it discourages more casual homebrew-core contributors from touching it. Not sure how that weighs against making sure the bottle block is easily readable.

Personally, I don't read the bottle block too often, so wouldn't mind seeing it moved downward. When I do, it's most often via `grep` to check if a formula has a specific bottle. Others may have different preferences, however.

Reply:
Since we seem to have some varying opinions here, do we want to solicit feedback from additional homebrew/core maintainers before moving in any particular direction?

Either way, I'll abide by whatever ends up being the final verdict here (updating this and the homebrew/core PR accordingly).

Reply:
In the interest of moving this toward a conclusion, I would appreciate feedback on this from @Homebrew/core.

Basically, it's necessary for `head` blocks to always appear before `livecheck` blocks (so `head` is defined before it's used in `Livecheck#checkable_urls` or `url :head`). The presented options to accomplish this are:

* Move `head` blocks to appear after `stable` blocks, both immediately before any `livecheck` block.
* Move `livecheck` blocks to appear after any `head` block (or `bottle` otherwise).

The expressed benefits/detriments of each approach can be found in the discussions above.

Reply:
Maybe you could assign reactions to the two options so you can collect opinions without needing to generate a lot of extra comments?

Reply:
What about storing the `:head` symbol and then getting the head url later in execution (after the formula has been loaded)?

Reply:
> What about storing the `:head` symbol and then getting the head url later in execution (after the formula has been loaded)?

I've previously considered making this change in `Livecheck#url` for other reasons. Namely, converting a URL symbol to the related URL string in the `#url` DSL method prevents us from being able to understand when a `livecheck` block uses a URL symbol (instead of a string literal).

The DSL methods that we implemented later aim to preserve the original value provided in the `livecheck` block. The idea behind this is that the `Livecheck` object should serve as an accurate representation of the data in a formula/cask's `livecheck` block and `brew livecheck` should handle conversions from symbols internally (e.g., from a `strategy` symbol to a strategy object).

Generally speaking, I'm in favor of modifying `Livecheck#url` to store the URL symbol and having `brew livecheck` convert from a symbol to the formula/cask URL internally. As you mentioned, this would sidestep the issue in this PR entirely but it also allows us to make the debug and JSON output richer (i.e., indicating when a `livecheck` block is using a URL symbol).

I think this may be the most amenable solution here (it's certainly the path of least resistance), so I'm going to go ahead and create a PR for this change. Thanks for bringing this up, @SeekingMeaning!